use output

patrickpatrickpatrick opened the pull request:
- **[WIP] Spike progressively enhanced file upload**
- **click label instead of input**
- **Add aria-hidden and tabindex -1 to input**
- **Add `multiple` Nunjucks parameter**
- **Tweaks to dropzone styles**
- **Remove setting aria-hidden**
- **Syncronise disabled state between input and button**
- **Add i18n parameters**
- **Write JavaScript functionality tests**
- **Fix component not rendering plural objects correctly**
- **Add Nunjucks parameter documentation**
- **Add tests for disable state syncronisation**
- **Show dropzone when dragover page and not input**
- **Change dragover to dragenter**
- **Add tests for initialisation errors specific to the component**
- **Make `FileUpload` extend `ConfigurableComponent`**
- **Format error message consistently**
- **Remove comments ignoring TypeScript or linting errors**
- **Remove unnecessary `if` statements**
- **Encapsulate lookup of the `<label>` element**
- **Remove unnecessary check for existence of the `files` attribute**
- **try using role button**

selfthinker commented:
I've done some testing with assistive technologies. Details are in the [testing spreadsheet](https://docs.google.com/spreadsheets/d/12qUQjVR19092-1t7S-8J3VDArczgs4XW9huxftRoSgc/edit?usp=sharing) (in the '16 Jan 2025' tab).
Here is a summary of my findings.

* The focus style of the fake button is currently not the same as our standard secondary button.
* There is no visible focus at all when using Windows High Contrast Mode or changing colours in Firefox.
* I can confirm that this version of the component works with Dragon.
* Most of the screen reader and browser combinations are pretty consistent, with a few exceptions.
* NVDA in Firefox never mentions the file name after a file has been selected. It only reads the label instead. I think that is because the screen reader focus is at the label for whatever reason. You can discover the file name when you navigate further.
* VoiceOver on macOS in Safari seems to read the hidden file input after selecting a file. (It says: "no file selected, [label], [hint], file upload button".) It usually says "no file selected" (which is what the native file input says in Safari when no file is selected), but sometimes it happened that it read the file name instead. Interestingly, the native file input also says "no file selected" after selecting a file, so it's not a bug that we have introduced. It's possible the JS gets the values correctly but that the values that Safari provides are just incorrect.
* It happened multiple times during testing with most screen readers that after selecting a file they either read "no file chosen" or the previously selected file. Could there be a race condition going on?

I suspect the only items from that list that are potentially gnarly are the two screen reader issues. We should definitely try to fix them for a bit. But the VO macOS / Safari issue has the lowest priority because of the pre-existing issue with the native input. It's worth trying to isolate the issue with NVDA to see where the issue comes from.
If we cannot fix either of these issues, it's worth googling if there already is a bug report about that behaviour and if there isn't, report it ourselves.

patrickpatrickpatrick commented:
Addressed a few comments:

- focus state (including the focus state box shadow, style when the button is focus and hovered and style when the button is active and focused) for pseudo button now same as secondary button
- added transparent outlines on focus for pseudo button for high contrast mode
- visually hidden `<span>, </span>` added between button and text to ensure gap between announcement of each aspect of the button

patrickpatrickpatrick commented:
New visually hidden field for reading the contents of the button, I'm hoping this will fix the screen reader issues.

romaricpascal reviewed (commented):
Almost there I think, that's neat! 🙌🏻 

Spotted an issue with Voice Over and Safari due to the setting of the ID on the `<button>` (good spot that we'd need to do that for the Error Summary links), but nothing unfixable, I think.

romaricpascal commented on packages/govuk-frontend/src/govuk/components/file-upload/file-upload.mjs:
```diff
@@ -68,6 +83,7 @@ export class FileUpload extends ConfigurableComponent {
     const $button = document.createElement('button')
     $button.classList.add('govuk-file-upload__button')
     $button.type = 'button'
+    $button.id = this.id
```
The announcement sounds better, but I think it's missing the 'Choose file' part, with VoiceOver and Safari, I'm only getting the label and the status text.

https://github.com/user-attachments/assets/407010df-1407-4762-a8d7-821d8eb6afa8

patrickpatrickpatrick replied:
This now announces the button text also

selfthinker commented:
I haven't done a full test yet, but just some findings from a quick initial test:

* The race condition still seems to be there.
* The hint (which is connected via `aria-describedby` to the original file input) is not connected yet and doesn't get read out when the focus is on the button.
* The label sometimes gets read out twice. I assume that's one from the `aria-label` and the other from the actual label. I need to test in other screen readers to confirm if that always happens, so safe to remove, or only sometimes. (I've so far only tested in NVDA/Chrome.)

selfthinker commented:
Because we're using `aria-label` here and `aria-labelledby` in #5639 and neither in #5640, I wanted to compare how these three methods behave when the page is translated automatically. I tested with the Google Translate feature that comes with the Chrome browser. Other translation features may behave very differently.

The good news is that it translate even the text in attributes (so, works with `aria-label`).
But it treats the text in the attribute differently than the one on the label. I assume that is because one stands on its own and the other is within a "sentence" with comma-separated other texts.
When I translated into German, the label said "Laden Sie Ihr Foto hoch" but in the `aria-label` as well as the hidden span for the `aria-labelledby` it said "Foto hochladen". Both mean the same thing but could be confusing for screen reader users who can see that they are different.

So, #5640 works the most consistently, although the others work well enough.

selfthinker commented:
Another thing to consider is users who use userstyles. That's a very small minority of users. While we don't necessarily need to ensure the page looks good with different CSS (or no CSS), it should at least be functional.

For those users the text on the button reads:

* "Choose fileNo file chosen" for this PR
* "Choose fileNo file chosenUpload your photo, Choose file, No file chosen" for #5639
* "Choose file, No file chosen" for #5640

I guess it's mostly still functional for all three, but getting weird.

romaricpascal commented on packages/govuk-frontend/src/govuk/components/file-upload/_index.scss:
```diff
@@ -97,3 +97,50 @@
+  background-color: transparent;
+}
+
+.govuk-file-upload:disabled + .govuk-file-upload__button {
+  pointer-events: none;
+}
+
+.govuk-file-upload:disabled + .govuk-file-upload__button .govuk-file-upload__pseudo-button {
+  opacity: (0.5);
+
+  &:hover {
+    background-color: govuk-colour("light-grey");
+    cursor: not-allowed;
+  }
+
+  &:active {
+    top: 0;
+    box-shadow: 0 $govuk-border-width-form-element 0 govuk-shade(govuk-colour("white"), 60%); // s0
+  }
+}
```
@patrickpatrickpatrick One for the PR about the styles, but given we set the `disabled` attribute on the `<button>`, we should directly target the `<button>` rather than use the state of the `<input>`.

romaricpascal reviewed (approved):
Ready to merge on the spike branch, the only remark I had likely affects the branch for the styling so we can update there 😊
